File: python/exercises/practice/unit_test_5cb1fe28_17f1_465d_a2f6_37cbc4b24a6a/distributed_rate_limiter/distributed_rate_limiter_test.py
import unittest
import time
from unittest.mock import patch

from distributed_rate_limiter import allow_request

class FakeTime:
    def __init__(self, start):
        self.current = start
    def time(self):
        return self.current
    def sleep(self, seconds):
        self.current += seconds

class DistributedRateLimiterTest(unittest.TestCase):
    def test_allow_request_within_limit(self):
        fake_time = FakeTime(1000)
        with patch('distributed_rate_limiter.time.time', side_effect=fake_time.time):
            rate_limit = 3
            time_window = 10
            user_id = "user1"
            api_endpoint = "/api/test"
            # Allow up to rate_limit requests at the same moment
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))
            # Next request should be blocked as it exceeds the rate limit
            self.assertFalse(allow_request(user_id, api_endpoint, rate_limit, time_window))

    def test_allow_request_window_reset(self):
        fake_time = FakeTime(2000)
        with patch('distributed_rate_limiter.time.time', side_effect=fake_time.time):
            rate_limit = 2
            time_window = 5
            user_id = "user2"
            api_endpoint = "/api/data"
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))
            # The next request should be blocked within the same time window
            self.assertFalse(allow_request(user_id, api_endpoint, rate_limit, time_window))
            # Advance time beyond the window to reset the count
            fake_time.sleep(time_window + 1)
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))

    def test_allow_request_different_users_endpoints(self):
        fake_time = FakeTime(3000)
        with patch('distributed_rate_limiter.time.time', side_effect=fake_time.time):
            rate_limit = 3
            time_window = 10
            # Test for user1 and endpoint /api/one
            self.assertTrue(allow_request("user1", "/api/one", rate_limit, time_window))
            self.assertTrue(allow_request("user1", "/api/one", rate_limit, time_window))
            self.assertTrue(allow_request("user1", "/api/one", rate_limit, time_window))
            self.assertFalse(allow_request("user1", "/api/one", rate_limit, time_window))
            # Test for same user with a different endpoint should have independent count
            self.assertTrue(allow_request("user1", "/api/two", rate_limit, time_window))
            self.assertTrue(allow_request("user1", "/api/two", rate_limit, time_window))
            self.assertTrue(allow_request("user1", "/api/two", rate_limit, time_window))
            self.assertFalse(allow_request("user1", "/api/two", rate_limit, time_window))
            # Test for different user on the same endpoint
            self.assertTrue(allow_request("user2", "/api/one", rate_limit, time_window))
            self.assertTrue(allow_request("user2", "/api/one", rate_limit, time_window))
            self.assertTrue(allow_request("user2", "/api/one", rate_limit, time_window))
            self.assertFalse(allow_request("user2", "/api/one", rate_limit, time_window))

    def test_allow_request_edge_time_window(self):
        fake_time = FakeTime(4000)
        with patch('distributed_rate_limiter.time.time', side_effect=fake_time.time):
            rate_limit = 2
            time_window = 5
            user_id = "edge"
            api_endpoint = "/edge/test"
            # First request at time 4000
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))
            fake_time.sleep(2)  # time becomes 4002
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))
            # Next request should be blocked because rate limit reached within time_window
            self.assertFalse(allow_request(user_id, api_endpoint, rate_limit, time_window))
            fake_time.sleep(3)  # time becomes 4005, edge of or just beyond the window
            # Now, the first request (at t=4000) falls outside the current window, so a new request is allowed
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))

    def test_allow_request_concurrent_simulation(self):
        # Simulate an interleaved/concurrent sequence of requests.
        fake_time = FakeTime(5000)
        with patch('distributed_rate_limiter.time.time', side_effect=fake_time.time):
            rate_limit = 5
            time_window = 10
            user_id = "concurrent"
            api_endpoint = "/api/concurrent"
            results = []
            # Simulate 5 requests, one per second
            for _ in range(5):
                results.append(allow_request(user_id, api_endpoint, rate_limit, time_window))
                fake_time.sleep(1)
            self.assertEqual(results, [True, True, True, True, True])
            # A burst request immediately after should be blocked
            self.assertFalse(allow_request(user_id, api_endpoint, rate_limit, time_window))
            # After the time window, requests should be allowed again
            fake_time.sleep(time_window)
            self.assertTrue(allow_request(user_id, api_endpoint, rate_limit, time_window))

if __name__ == '__main__':
    unittest.main()

File: cpp/exercises/practice/unit_test_e50cd626_89f3_4e19_9a7c_cd4a94675a70/distributed_rate_limiter/distributed_rate_limiter_test.cpp
#include "distributed_rate_limiter.h"
#include "catch.hpp"

TEST_CASE("Single request within limit", "[RateLimiter]") {
    RateLimiter limiter;
    limiter.setRateLimit("user1", 5, 10); // 5 requests per 10 seconds

    // First 5 requests should be allowed.
    for (int i = 0; i < 5; ++i) {
        REQUIRE(limiter.allowRequest("user1", 100));
    }
}

TEST_CASE("Exceeding rate limit", "[RateLimiter]") {
    RateLimiter limiter;
    limiter.setRateLimit("user2", 3, 10); // 3 requests per 10 seconds
    int timestamp = 200;

    // 3 requests allowed.
    REQUIRE(limiter.allowRequest("user2", timestamp));
    REQUIRE(limiter.allowRequest("user2", timestamp));
    REQUIRE(limiter.allowRequest("user2", timestamp));
    // 4th request within same window should be rejected.
    REQUIRE_FALSE(limiter.allowRequest("user2", timestamp));
}

TEST_CASE("Window expiration resets limit", "[RateLimiter]") {
    RateLimiter limiter;
    limiter.setRateLimit("user3", 2, 5); // 2 requests per 5 seconds

    int t1 = 300;
    // Consume the rate limit.
    REQUIRE(limiter.allowRequest("user3", t1));
    REQUIRE(limiter.allowRequest("user3", t1));
    REQUIRE_FALSE(limiter.allowRequest("user3", t1));

    // After time window expires, the limit should reset.
    int t2 = t1 + 6;
    REQUIRE(limiter.allowRequest("user3", t2));
    REQUIRE(limiter.allowRequest("user3", t2));
    REQUIRE_FALSE(limiter.allowRequest("user3", t2));
}

TEST_CASE("Multiple users", "[RateLimiter]") {
    RateLimiter limiter;
    limiter.setRateLimit("userA", 2, 10); // User A: 2 req per 10 sec
    limiter.setRateLimit("userB", 4, 10); // User B: 4 req per 10 sec

    int t = 400;
    // Test for userA.
    REQUIRE(limiter.allowRequest("userA", t));
    REQUIRE(limiter.allowRequest("userA", t));
    REQUIRE_FALSE(limiter.allowRequest("userA", t));

    // Test for userB.
    for (int i = 0; i < 4; ++i) {
        REQUIRE(limiter.allowRequest("userB", t));
    }
    REQUIRE_FALSE(limiter.allowRequest("userB", t));
}

TEST_CASE("Consecutive windows with expiring requests", "[RateLimiter]") {
    RateLimiter limiter;
    limiter.setRateLimit("userX", 3, 10); // 3 requests per 10 seconds

    // Issue requests at different timestamps.
    REQUIRE(limiter.allowRequest("userX", 0));  // first request at t=0
    REQUIRE(limiter.allowRequest("userX", 5));  // second request at t=5
    REQUIRE(limiter.allowRequest("userX", 9));  // third request at t=9
    // At t=10, the request at t=0 should expire.
    REQUIRE(limiter.allowRequest("userX", 10));
    // Within window [5,10], there are now 3 requests; next should be rejected.
    REQUIRE_FALSE(limiter.allowRequest("userX", 10));
}

TEST_CASE("Rapid succession and random timing", "[RateLimiter]") {
    RateLimiter limiter;
    limiter.setRateLimit("user_random", 5, 10); // 5 requests per 10 sec

    int start = 500;
    int allowedCount = 0;
    // Simulate requests at timestamps start, start+1, ..., start+9.
    for (int i = 0; i < 10; ++i) {
        if (limiter.allowRequest("user_random", start + i)) {
            allowedCount++;
        }
    }
    // Ensure that within any 10 second window, the allowed count does not exceed 5.
    REQUIRE(allowedCount <= 5);
}